testing/integration-testing-hooks.ts

```diff
@@ -1,92 +1,370 @@
-const mongodbRunnerBefore = require('mongodb-runner/mocha/before');
-const mongodbRunnerAfter = require('mongodb-runner/mocha/after');
-const semver = require('semver');
-const { MongoClient } = require('mongodb');
+import semver from 'semver';
+// Installing @types/mongodb at the lerna root breaks some of the tests...
+// Yikes. Stick with ts-ignore for now.
+// @ts-ignore
+import { MongoClient } from 'mongodb';
+import { promisify } from 'util';
+import child_process from 'child_process';
+import crypto from 'crypto';
+import { promises as fs } from 'fs';
+import path from 'path';
+import rimraf from 'rimraf';
+import { URL } from 'url';
+import which from 'which';
+import { once } from 'events';
+import { downloadMongoDb } from './download-mongodb';
 
-export const LOCAL_INSTANCE_PORT = 27018;
-export const LOCAL_INSTANCE_HOST = 'localhost';
+const execFile = promisify(child_process.execFile);
+
+// Return the stat results or, if the file does not exist, `undefined`.
+async function statIfExists(path: string): Promise<ReturnType<typeof fs.stat> | undefined> {
+  try {
+    return await fs.stat(path);
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      return undefined;
+    }
+    throw err;
+  }
+}
+
+// Return the path to the temporary directory and ensure that it exists.
+async function getTmpdir(): Promise<string> {
+  const tmpdir = path.resolve(__dirname, '..', 'tmp');
+  await fs.mkdir(tmpdir, { recursive: true });
+  return tmpdir;
+}
+
+let mlaunchPath: { exec: string, env: Record<string,string> } | undefined;
+async function getMlaunchPath(): Promise<{ exec: string, env: Record<string,string> }> {
+  const tmpdir = await getTmpdir();
+  if (mlaunchPath !== undefined) {
+    return mlaunchPath;
+  }
+
+  try {
+    // If `mlaunch` is already in the PATH: Great, we're done.
+    return mlaunchPath = { exec: await which('mlaunch'), env: {} };
+  } catch {}
+
+  let python = '';
+  try {
+    await which('python3');
+    python = 'python3';
+  } catch {
+    try {
+      // Fun fact on the side: Python 2.x writes the version to stderr,
+      // Python 3.x writes to stdout.
+      const { stdout } = await execFile('python', ['-V']);
+      if (stdout.includes('Python 3')) {
+        python = 'python';
+      }
+    } catch {
+      const pythonEnv = process.env.PYTHON;
+      if (pythonEnv) {
+        const { stdout } = await execFile(pythonEnv as string, ['-V']);
+        if (stdout.includes('Python 3')) {
+          python = pythonEnv as string;
+        }
+      }
+    }
+  }
+  if (!python) {
+    throw new Error('Could not find Python 3.x installation, install mlaunch manually');
+  }
+
+  // Install mlaunch, preferably locally and otherwise attempt to do so globally.
+  try {
+    await execFile('pip3', ['install', '--target', tmpdir, 'mtools[mlaunch]']);
```

Comment:
I'm afraid just installing that is not enough. I just tried it on my macOS with basically a really plain setup (no PIP packages or anything so far). When trying to run the tests I was presented with `ImportErrors` for the following packages:
- `psutil`
- `pymongo`

After manually installing these packages using PIP it worked like a charm 😄 So we should add them to the list of installed packages here.

Reply:
@rose-m This is very odd, pip should install dependencies automatically … I’ll try to look into that

Reply:
@rose-m I could not really reproduce this, cleared out the installed python modules on my mac and ran the tests succesfully … so I’m not sure how to progress here :confused: We can add them, but it’s odd that that would be necessary. Can you try to reproduce the problem, and maybe add a `console.log()` here that prints whatever `pip3` logged (just wrapping this line in it should be enough)?

Reply:
Yup, I'll give it another try!

Reply:
Okay I cannot reproduce anymore... it's clear that PIP automatically downloads and installs the dependencies as you said... Tried cleaning everything up and removing caches - works fine... So I'd say we can mark this comment as resolved then. Sorry 🤷